File: app/src/main/java/com/futsch1/medtimer/database/ReminderEvent.java
package com.futsch1.medtimer.database;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Index;
import androidx.room.PrimaryKey;

import com.google.gson.annotations.Expose;

import java.util.List;
import java.util.Objects;

@Entity(indices = {@Index(value = "reminderId"), @Index("remindedTimestamp")})
public class ReminderEvent {
    @PrimaryKey(autoGenerate = true)
    public int reminderEventId;
    @Expose
    public String medicineName;
    @Expose
    public String amount;
    @ColumnInfo(defaultValue = "0")
    @Expose
    public int color;
    @ColumnInfo(defaultValue = "false")
    @Expose
    public boolean useColor;
    @Expose
    public ReminderStatus status;
    @Expose
    public long remindedTimestamp;
    @Expose
    public long processedTimestamp;
    @Expose
    public int reminderId;
    @ColumnInfo(defaultValue = "0")
    public int notificationId;
    @ColumnInfo(defaultValue = "0")
    @Expose
    public int iconId;
    @ColumnInfo(defaultValue = "0")
    public int remainingRepeats;
    @ColumnInfo(defaultValue = "false")
    public boolean stockHandled;
    @ColumnInfo(defaultValue = "false")
    public boolean askForAmount;
    @ColumnInfo(defaultValue = "[]")
    @Expose
    public List<String> tags;
    @ColumnInfo(defaultValue = "0")
    @Expose
    public int lastIntervalReminderTimeInMinutes;
    @ColumnInfo(defaultValue = "")
    @Expose
    public String notes;

    public ReminderEvent() {
        medicineName = "";
        amount = "";
        color = 0;
        useColor = false;
        status = ReminderStatus.RAISED;
        remindedTimestamp = 0;
        processedTimestamp = 0;
        reminderId = 0;
        notificationId = 0;
        iconId = 0;
        remainingRepeats = 0;
        stockHandled = false;
        askForAmount = false;
        tags = List.of();
        lastIntervalReminderTimeInMinutes = 0;
        notes = "";
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        return membersEqual((ReminderEvent) o);
    }

    @Override
    public int hashCode() {
        return Objects.hash(reminderEventId, medicineName, amount, color, useColor, status, remindedTimestamp, processedTimestamp, reminderId, notificationId, iconId, remainingRepeats, stockHandled, askForAmount, tags, lastIntervalReminderTimeInMinutes, notes);
    }

    private boolean membersEqual(ReminderEvent o) {
        return reminderEventId == o.reminderEventId &&
                Objects.equals(medicineName, o.medicineName) &&
                Objects.equals(amount, o.amount) &&
                color == o.color &&
                useColor == o.useColor &&
                status == o.status &&
                remindedTimestamp == o.remindedTimestamp &&
                processedTimestamp == o.processedTimestamp &&
                reminderId == o.reminderId &&
                notificationId == o.notificationId &&
                iconId == o.iconId &&
                remainingRepeats == o.remainingRepeats &&
                stockHandled == o.stockHandled &&
                askForAmount == o.askForAmount &&
                Objects.equals(tags, o.tags) &&
                lastIntervalReminderTimeInMinutes == o.lastIntervalReminderTimeInMinutes &&
                Objects.equals(notes, o.notes);
    }

    public enum ReminderStatus {
        RAISED,
        TAKEN,
        SKIPPED,
        DELETED
    }
}

File: app/src/main/java/com/futsch1/medtimer/database/ReminderEvent.kt
package com.futsch1.medtimer.database

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import com.futsch1.medtimer.database.Reminder.ReminderType
import com.google.gson.annotations.Expose
import java.util.Objects

@Entity(indices = [Index("reminderId"), Index("remindedTimestamp")])
class ReminderEvent {
    @JvmField
    @PrimaryKey(autoGenerate = true)
    var reminderEventId: Int = 0

    @JvmField
    @Expose
    var medicineName: String = ""

    @JvmField
    @Expose
    var amount: String = ""

    @JvmField
    @ColumnInfo(defaultValue = "0")
    @Expose
    var color: Int = 0

    @JvmField
    @ColumnInfo(defaultValue = "false")
    @Expose
    var useColor: Boolean = false

    @JvmField
    @Expose
    var status: ReminderStatus = ReminderStatus.RAISED

    @JvmField
    @Expose
    var remindedTimestamp: Long = 0

    @JvmField
    @Expose
    var processedTimestamp: Long = 0

    @JvmField
    @Expose
    var reminderId: Int = 0

    @JvmField
    @ColumnInfo(defaultValue = "0")
    var notificationId: Int = 0

    @JvmField
    @ColumnInfo(defaultValue = "0")
    @Expose
    var iconId: Int = 0

    @JvmField
    @ColumnInfo(defaultValue = "0")
    var remainingRepeats: Int = 0

    @JvmField
    @ColumnInfo(defaultValue = "false")
    var stockHandled: Boolean = false

    @JvmField
    @ColumnInfo(defaultValue = "false")
    var askForAmount: Boolean = false

    @JvmField
    @ColumnInfo(defaultValue = "[]")
    @Expose
    var tags: List<String> = listOf()

    @JvmField
    @ColumnInfo(defaultValue = "0")
    @Expose
    var lastIntervalReminderTimeInMinutes: Int = 0

    @JvmField
    @ColumnInfo(defaultValue = "")
    @Expose
    var notes: String = ""

    @JvmField
    @ColumnInfo(defaultValue = "TIME_BASED")
    @Expose
    var reminderType: ReminderType = ReminderType.TIME_BASED

    val isOutOfStockOrExpirationReminder: Boolean
        get() = reminderType == ReminderType.OUT_OF_STOCK || reminderType == ReminderType.EXPIRATION_DATE

    override fun equals(other: Any?): Boolean {
        if (other == null || javaClass != other.javaClass) return false
        return membersEqual(other as ReminderEvent)
    }

    override fun hashCode(): Int {
        return Objects.hash(
            reminderEventId,
            medicineName,
            amount,
            color,
            useColor,
            status,
            remindedTimestamp,
            processedTimestamp,
            reminderId,
            notificationId,
            iconId,
            remainingRepeats,
            stockHandled,
            askForAmount,
            tags,
            lastIntervalReminderTimeInMinutes,
            notes,
            reminderType
        )
    }

    private fun membersEqual(o: ReminderEvent): Boolean {
        return reminderEventId == o.reminderEventId &&
                medicineName == o.medicineName &&
                amount == o.amount && color == o.color && useColor == o.useColor && status == o.status && remindedTimestamp == o.remindedTimestamp && processedTimestamp == o.processedTimestamp && reminderId == o.reminderId && notificationId == o.notificationId && iconId == o.iconId && remainingRepeats == o.remainingRepeats && stockHandled == o.stockHandled && askForAmount == o.askForAmount &&
                tags == o.tags && lastIntervalReminderTimeInMinutes == o.lastIntervalReminderTimeInMinutes &&
                notes == o.notes && reminderType == o.reminderType
    }

    enum class ReminderStatus {
        RAISED,
        TAKEN,
        SKIPPED,
        DELETED,
        ACKNOWLEDGED
    }

    companion object {
        val allStatusValues: List<ReminderStatus> = ReminderStatus.entries
        val statusValuesWithoutDelete: List<ReminderStatus> = ReminderStatus.entries.filterNot { it == ReminderStatus.DELETED }
        val statusValuesWithoutAcknowledgedAndDeleted: List<ReminderStatus> =
            ReminderStatus.entries.filterNot { it == ReminderStatus.ACKNOWLEDGED || it == ReminderStatus.DELETED }
    }
}
